lib/t/matches.ex

```diff
@@ -895,6 +958,19 @@ defmodule T.Matches do
         %{contacts: contacts}
       )
 
+    match_event = %MatchEvent{
+      timestamp: now,
+      match_id: match_id,
+      event: "contact_offer"
+    }
+
+    interaction = %Interaction{
+      data: %{"type" => "contact_offer", "contacts" => contacts},
```

Comment:
`contacts` should become just a single `contact` now (https://github.com/getsince/test3/pull/540/files#diff-431ecb538ba81baef40c63a16664a4e648660f51a14f8228b916130fe0ba2237R944)

I don't think it's needed, but if you want it to avoid breaking changes, we can still accept multiple contacts, but insert them one by one

Reply:
>  but if you want it to avoid breaking changes

What breaking change does using `contacts` instead of `contact` introduce?

Reply:
> What breaking change does using contacts instead of contact introduce?

I meant switching to single `contact` from current `contacts` will introduce a breaking change

Reply:
I'm still using `contacts`, where is `contact` coming from?

Reply:
yes, I am suggesting switching to `contact` since `contacts` suited our previous model (we rendered all sent contacts at once in a popup), but do not suit our current model (where each contact is a separate chat item)

Reply:
what stops you from rendering contacts as separate chat items?

Reply:
order has to be preserved, i.e.:
i send you "telegram" contact
you send me voice message
i send you "whatsapp" contact

Reply:
order would be preserved

Reply:
the correct order will be:
1. "telegram" contact
2. voice message
3. "whatsapp" contact
I do not see how I will be able to preserve this order with current backend code, maybe I am missing something?

Reply:
```swift
let interactions: [MatchInteraction] = [
  .contactOffer(contacts: Contacts(telegram: "...")),
  .voicemail(...),
  .contactOffer(contacts: Contacts(whatsapp: "...")),
]
```

Reply:
ah, right, I see where you are coming from now

but current frontend unfortunately doesn't work that way: https://github.com/getsince/test5/blob/rails-fork/Since/Feeds/Normal%20Feed/NormalFeedChannel.swift#L266
so it will effectively always append to contacts list making it:
1) "telegram" contact
2) voice message
3) "telegram" contact + "whatsapp" contact

Reply:
right, but if we enforce "one contact" at a time things should be named properly + for contact type `interaction` rendering frontend will need to pick just one contact out of these

Reply:
> we enforce "one contact" at a time

how?

Reply:
> How?

in current version of frontend there will be only one contact sent at a time

> No matter what approach you pick, you still can do it with both representations, since they are equivalent.

if there is only one interaction - i get exactly the first contact I am trying (e.g. if dict["telegram"] else if dict["whatsapp"] etc.); if there are many - I will get the first one I am asking for (which is probably still fine since the update curve is pretty steep + not that many users will send multiple contacts using old version)

Reply:
```swift
struct Contact {
  let type: String
  let value: String
}

enum MatchInteraction {
  case contact(Contact)
  case voicemail
}

let interactions: [[String: Any]] = [
  ["type": "contact_offer", "contacts": ["telegram": "@telegram"]],
  ["type": "voicemail"],
  ["type": "contact_offer",  "contacts": ["whatsapp": "@whatsapp", "telegram": "@telegram"]]
]

func decode(_ json: [[String: Any]]) -> [MatchInteraction] {
  var seenContactsFromMe = MapSet<Contact>()
  json.compactMap { raw in
    guard let type = raw["type"] as? String else { return nil }
    
    switch type {
    case "contact_offer": 
    }
  }
}
```